File: CivitasJava/src/civitas/SorpresaPorJugador.java
package civitas;

import java.util.ArrayList;

public class SorpresaPorJugador extends Sorpresa{

    SorpresaPorJugador(int valor, String texto){
        super(texto);
        this.valor=valor;
    }

    @Override
    void aplicarAJugador(int actual, ArrayList<Jugador> todos){
        if(jugadorCorrecto(actual,todos)){
            informe(actual,todos);
            
            int valor_pagar= valor*-1;
            SorpresaPagarCobrar pagar=new SorpresaPagarCobrar(valor_pagar, "Pierdes 400 euros");
            for(int i=0;i<todos.size();i++){
                if(i!=actual){
                    pagar.aplicarAJugador(i, todos);
                }
            }
            
            int valor_cobrar= (valor*(todos.size()-1));
            SorpresaPagarCobrar cobrar=new SorpresaPagarCobrar(valor_cobrar, "Recibes todo el dinero");
            cobrar.aplicarAJugador(actual, todos);
        }
    }
}

File: CivitasRuby/lib/sorpresa_por_jugador.rb
#encoding:utf-8

require_relative 'sorpresa'
require_relative 'jugador'
require_relative 'sorpresa_pagar_cobrar'
require_relative 'tablero'
require_relative 'casilla'
require_relative 'mazo_sorpresas'
require 'sorpresa.rb'

module Civitas
    class SorpresaPorJugador < Sorpresa 
        def initialize(valor, texto)
            super(texto)
            @valor=valor
        end

        def aplicar_a_jugador(actual, todos)
            if(jugador_correcto(actual,todos))
                informe(actual, todos)
                
                valor_pagar=@valor*-1
                pagar=SorpresaPagarCobrar.new(valor_pagar, "Pierdes 400 euros")
                for i in 0..(todos.length-1)
                    if(i!=actual)
                        pagar.aplicar_a_jugador(i,todos)
                    end
                end
                
                valor_cobrar=@valor*(todos.length-1)
                cobrar = SorpresaPagarCobrar.new(valor_cobrar, "Recibes todo el dinero")
                cobrar.aplicar_a_jugador(actual,todos)
            end
        end

        public_class_method :new
    end
end
            